### PR TITLE
Adds support for fetching alerts from custom alert index in Get Alerts transport layer

### PR DESCRIPTION
**Pull request by eirsep**
Adds support for fetching alerts from custom alert index in Get Alerts transport layer

Signed-off-by: Surya Sashank Nistala <snistala@amazon.com>

Related Issues:
#562 #522

### CODE REVIEW

File: alerting/src/main/kotlin/org/opensearch/alerting/action/GetAlertsRequest.kt [hunk 1]
```diff
@@ -17,25 +17,29 @@ class GetAlertsRequest : ActionRequest {
...
     val monitorId: String?
+    val alertIndex: String?
 
     constructor(
         table: Table,
         severityLevel: String,
         alertState: String,
-        monitorId: String?
+        monitorId: String?,
+        alertIndex: String?,
```
Comment (getsaurabh02) [inline alerting/src/main/kotlin/org/opensearch/alerting/action/GetAlertsRequest.kt:27]: Do we need to add the version check for backward compatibility?

File: alerting/src/main/kotlin/org/opensearch/alerting/transport/TransportGetAlertsAction.kt [hunk 2]
```diff
@@ -103,39 +107,75 @@ class TransportGetAlertsAction @Inject constructor(
...
+            alertIndex = getAlertsRequest.alertIndex
+            getAlerts(searchSourceBuilder, alertIndex, actionListener, user)
+        } else if (getAlertsRequest.monitorId != null) {
+            transportGetMonitorAction.execute(
+                GetMonitorRequest(
+                    getAlertsRequest.monitorId,
+                    -3L,
+                    RestRequest.Method.GET,
+                    FetchSourceContext.FETCH_SOURCE
+                ),
```
Comment (getsaurabh02) [inline alerting/src/main/kotlin/org/opensearch/alerting/transport/TransportGetAlertsAction.kt:121]: should we move this to a common util and try to re-use across places wherever we need to fetch the monitor config?

File: alerting/src/main/kotlin/org/opensearch/alerting/transport/TransportGetAlertsAction.kt [hunk 3]
```diff
@@ -103,39 +107,75 @@ class TransportGetAlertsAction @Inject constructor(
...
+                        }
+                    }
+
+                    override fun onFailure(t: Exception) {
+                        actionListener.onFailure(t)
+                    }
+                }
+            )
+        } else {
+            getAlerts(searchSourceBuilder, alertIndex, actionListener, user)
```
Comment (getsaurabh02) [inline alerting/src/main/kotlin/org/opensearch/alerting/transport/TransportGetAlertsAction.kt:136]: ``getAlerts`` is being called 3 times here in this block. We should first resolve the index name and then make the final call.

File: alerting/src/main/kotlin/org/opensearch/alerting/transport/TransportGetAlertsAction.kt [hunk 4]
```diff
@@ -103,39 +107,75 @@ class TransportGetAlertsAction @Inject constructor(
...
+                object : ActionListener<GetMonitorResponse> {
+                    override fun onResponse(getMonitorResponse: GetMonitorResponse) {
+                        if (getMonitorResponse.monitor != null) {
+                            alertIndex = getMonitorResponse.monitor!!.dataSources.alertsIndex
+                            getAlerts(searchSourceBuilder, alertIndex, actionListener, user)
+                        }
+                    }
+
+                    override fun onFailure(t: Exception) {
+                        actionListener.onFailure(t)
```
Comment (getsaurabh02) [inline alerting/src/main/kotlin/org/opensearch/alerting/transport/TransportGetAlertsAction.kt:131]: During failure, should we rather log the message and fallback to default index name?

File: alerting/src/test/kotlin/org/opensearch/alerting/MonitorDataSourcesIT.kt [hunk 5]
```diff
@@ -44,6 +47,24 @@ class MonitorDataSourcesIT : AlertingSingleNodeTestCase() {
...
+        try {
+            client()
+                .execute(GetAlertsAction.INSTANCE, GetAlertsRequest(table, "ALL", "ALL", null, "wrong_alert_index"))
+                .get()
+            fail()
+        } catch (e: Exception) {
+        }
     }
 
     fun `test execute monitor with custom alerts index`() {
```
Comment (getsaurabh02) [inline alerting/src/test/kotlin/org/opensearch/alerting/MonitorDataSourcesIT.kt:70]: Lets add test for failure scenarios, such as index not present, monitor config-resolution failures?

**Review by getsaurabh02 — CHANGES_REQUESTED** [review]

File: alerting/src/main/kotlin/org/opensearch/alerting/transport/TransportGetAlertsAction.kt (hunk 4 above)
Comment (eirsep) [reply alerting/src/main/kotlin/org/opensearch/alerting/transport/TransportGetAlertsAction.kt:131]: +1 on logging message. Will do that. This was existing code which was moved to a method hence overlooked its correctness.

But if the user passes an incorrect index name I don't think we should fallback to default as we might give them misleading info. They might assume we are surfacing alerts from the custom index name that they passed. It's better to throw an error which might make them evaluate what went wrong. Silent failure might be counterproductive.

File: alerting/src/main/kotlin/org/opensearch/alerting/transport/TransportGetAlertsAction.kt (hunk 2 above)
Comment (eirsep) [reply alerting/src/main/kotlin/org/opensearch/alerting/transport/TransportGetAlertsAction.kt:121]: let's get this PR in along with #563 and I will do that as a follow up.

File: alerting/src/main/kotlin/org/opensearch/alerting/action/GetAlertsRequest.kt (hunk 1 above)
Comment (eirsep) [reply alerting/src/main/kotlin/org/opensearch/alerting/action/GetAlertsRequest.kt:27]: created a task to track this.

Comment (eirsep) [reply alerting/src/main/kotlin/org/opensearch/alerting/action/GetAlertsRequest.kt:27]: this question applies to changes that we have been making in the overall Datasources purview. WIll revisit during refactoring models to common utils

File: alerting/src/test/kotlin/org/opensearch/alerting/MonitorDataSourcesIT.kt (hunk 5 above)
Comment (eirsep) [reply alerting/src/test/kotlin/org/opensearch/alerting/MonitorDataSourcesIT.kt:70]: have added check to verify index not found scenario.

File: alerting/src/main/kotlin/org/opensearch/alerting/transport/TransportGetAlertsAction.kt (hunk 4 above)
Comment (getsaurabh02) [reply alerting/src/main/kotlin/org/opensearch/alerting/transport/TransportGetAlertsAction.kt]: I am thinking about the stale config information being treated same everywhere. But yeah we can revisit this on the whole.

**Review by getsaurabh02 — APPROVED** [review]

**Review by sbcd90 — APPROVED** [review]

**Pull request merged**